python/paddle/distributed/auto_parallel/static/tuner/pir_rule_based_tuner.py

```diff
@@ -0,0 +1,1487 @@
+        out_linear_dist_infos = MpDistInfos("row")
+        up_linear_dist_infos = MpDistInfos("column")
+        down_linear_dist_infos = MpDistInfos("row")
+
+        # # # build ops dist infos # # #
+        ops_dist_infos = {
+            (22,): qkv_linear_dist_infos,
+            (23,): qkv_linear_dist_infos,
+            (24,): qkv_linear_dist_infos,
+            (90,): out_linear_dist_infos,
+            (100,): up_linear_dist_infos,
+            (101,): up_linear_dist_infos,
+            (103,): down_linear_dist_infos,
+        }
+        self.ops_dist_infos = ops_dist_infos
+
+    def rms_norm(self, hidden_states, norm_weight):
+        variance_epsilon = 1e-6
+        with paddle.amp.auto_cast(False):
+            hidden_states = hidden_states.astype("float32")
+            variance = hidden_states.pow(2).mean(-1, keepdim=True)
+            hidden_states = (
+                paddle.rsqrt(variance + variance_epsilon) * hidden_states
+            )
+            out = hidden_states * norm_weight
+        return out
+
+    def attention(
+        self,
+        hidden_states,
+        q_weight,
+        k_weight,
+        v_weight,
+        out_weight,
+        cos_cached,
+        sin_cached,
+        position_ids,
+        attention_mask,
+        num_heads,
+        head_size,
+        seq_length,
+    ):
+        # qkv (not fused)
+        q = paddle.matmul(hidden_states, q_weight)
+        k = paddle.matmul(hidden_states, k_weight)
+        v = paddle.matmul(hidden_states, v_weight)
+        target_q_shape = [0, 0, num_heads, head_size]
+        target_kv_shape = [0, 0, num_heads, head_size]
+        q = q.reshape(shape=target_q_shape)
+        k = k.reshape(shape=target_kv_shape)
+        v = v.reshape(shape=target_kv_shape)
+
+        # rope_emb
+        cos = cos_cached[:, :seq_length, :, :]
+        sin = sin_cached[:, :seq_length, :, :]
+
+        # apply_rotary_pos_emb
+        if position_ids is None:
+            cos = cos[:, : q.shape[1], :, :]  # [bs, seq_len, 1, dim]
+            sin = sin[:, : q.shape[1], :, :]  # [bs, seq_len, 1, dim]
+        else:
+            cos = cos.squeeze(axis=[0, 2])  # [seq_len, dim]
+            sin = sin.squeeze(axis=[0, 2])  # [seq_len, dim]
+            cos = cos[position_ids].unsqueeze(2)  # [bs, seq_len, 1, dim]
+            sin = sin[position_ids].unsqueeze(2)  # [bs, seq_len, 1, dim]
+        q_embed = (q * cos) + (self.rotate_half(q) * sin)
+        k_embed = (k * cos) + (self.rotate_half(k) * sin)
+
+        # scale_dot_product
+        tmp = self.scale_dot_product_attention(
+            q_embed, k_embed, v, attention_mask
+        )
+
+        # out_linear
+        output = paddle.matmul(tmp, out_weight)
+        return output
+
+    def rotate_half(self, x):
+        x1 = x[..., : x.shape[-1] // 2]
+        x2 = x[..., x.shape[-1] // 2 :]
+        return paddle.concat([-x2, x1], axis=-1)  # shape is the same as x
+
+    def scale_dot_product_attention(
```

Comment:
The code of this function is basically the same as the code of the function in `PIRScaleDotProductPattern` above, and `rotate_half`, `mlp`, etc. also have the same problem. complex networks should be a combination of basic networks, rather than a re implementation.

Reply:
done